File: ollama/codigo-final/fasta/codegemma-7b/cot/temp_0/fasta.cpp
#include <iostream>
#include <string>
#include <random>

using namespace std;

const string ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA"
                 "TCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACT"
                 "AAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAG"
                 "GCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCG"
                 "CCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

const double IUB_probs[] = {0.27, 0.12, 0.12, 0.27, 0.02};
const double Homo_probs[] = {0.3029549426680, 0.1979883004921, 0.1975473066391, 0.3015094502008};

const int IM = 139968;
const int IA = 3877;
const int IC = 29573;
const int Seed = 42;

int main(int argc, char* argv[]) {
  if (argc != 2) {
    cerr << "Usage: fasta N" << endl;
    return 1;
  }

  int N = atoi(argv[1]);
  if (N <= 0) {
    return 0;
  }

  random_device rd;
  mt19937 gen(rd());
  uniform_real_distribution<> dis(0.0, 1.0);

  for (int n : {N * 2, N * 3, N * 5}) {
    cout << ">ONE Homo sapiens alu" << endl;
    for (int i = 0; i < n; i++) {
      cout << ALU[i % ALU.size()];
      if (i % 60 == 59) {
        cout << endl;
      }
    }
    cout << endl;

    cout << ">TWO IUB ambiguity codes" << endl;
    for (int i = 0; i < n; i++) {
      double r = dis(gen);
      int index = 0;
      while (r > 0) {
        r -= IUB_probs[index++];
      }
      cout << "ACGTBNDHRVWSY"[index - 1];
      if (i % 60 == 59) {
        cout << endl;
      }
    }
    cout << endl;

    cout << ">THREE Homo sapiens frequency" << endl;
    for (int i = 0; i < n; i++) {
      double r = dis(gen);
      int index = 0;
      while (r > 0) {
        r -= Homo_probs[index++];
      }
      cout << "ACGT"[index - 1];
      if (i % 60 == 59) {
        cout << endl;
      }
    }
    cout << endl;
  }

  return 0;
}

File: ollama/codigo-final/fasta/codegemma-7b/cot/temp_0/fasta.py
import sys

# LCG parameters
IM = 139968
IA = 3877
IC = 29573
Seed = 42
Max = 1.0

# ALU sequence
ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA" \
      "TCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACT" \
      "AAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAG" \
      "GCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCG" \
      "CCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

# IUB alphabet probabilities
IUB_probs = [0.27, 0.12, 0.12, 0.27] + [0.02] * 8

# Homo sapiens alphabet probabilities
Homo_probs = [0.3029549426680, 0.1979883004921, 0.1975473066391, 0.3015094502008]

def generate_random_number(seed):
    """Linear Congruential Generator"""
    Seed = (IA * Seed + IC) % IM
    return Seed / IM

def select_nucleotide(cumulative_probs, random_num):
    """Weighted selection based on cumulative probabilities"""
    for i, prob in enumerate(cumulative_probs):
        if random_num < prob:
            return i

def generate_sequence(alphabet, n_chars):
    """Generate a sequence of nucleotides"""
    sequence = ""
    for _ in range(n_chars):
        random_num = generate_random_number(Seed)
        index = select_nucleotide(alphabet, random_num)
        sequence += alphabet[index]
    return sequence

def main():
    # Parse command-line argument
    try:
        N = int(sys.argv[1])
    except ValueError:
        print("Invalid command-line argument. Please provide a valid integer for N.")
        sys.exit(1)

    # Generate sequences
    alu_sequence = ALU * (N // 2)
    iub_sequence = generate_sequence(IUB_probs, N * 3)
    homo_sequence = generate_sequence(Homo_probs, N * 5)

    # Write sequences to file
    with open("output.txt", "w") as f:
        f.write(">ONE Homo sapiens alu\n")
        f.write(alu_sequence + "\n")
        f.write(">TWO IUB ambiguity codes\n")
        f.write(iub_sequence + "\n")
        f.write(">THREE Homo sapiens frequency\n")
        f.write(homo_sequence + "\n")

if __name__ == "__main__":
    main()
